imap_processing/glows/l1/glows_l1a_data.py

```diff
@@ -0,0 +1,411 @@
+        data class in the sequence.
+    block_header: dict
+        Block header data structure containing versioning, sequence, and file names
+    de_data: bytearray
+        Bytearray of raw DirectEvent data, which is converted into direct_events
+    most_recent_seq: int
+        The most recent sequence added to the L1A dataclass - for counting gaps
+    missing_seq: list[int]
+        Any missing sequence counts in the data. Should be an empty array in normal
+        operation
+    status_data: dict
+        Dictionary of the status data, with names defined by DataEverySecond. Defaults
+        to None.
+    direct_events: list[DirectEvent]
+        List of DirectEvent objects, which is created when the final level 0 packet in
+        the sequence is added to de_data. Defaults to None.
+
+    Methods
+    -------
+    append
+        Add another Level0 instance
+    """
+
+    l0: DirectEventL0
+    block_header: dict
+    de_data: bytearray
+    most_recent_seq: int
+    missing_seq: list[int]
+    status_data: dict = None
+    direct_events: list[DirectEvent] = None
+
+    def __init__(self, level0: DirectEventL0):
+        self.l0 = level0
+        self.most_recent_seq = self.l0.SEQ
+        self.de_data = bytearray(level0.DE_DATA)
+        self.missing_seq = []
+
+        self.block_header = {
+            "ground_software_version": version,
+            "pkts_file_name": self.l0.packet_file_name,
+            # note: packet number is seq_count (per apid!) field in CCSDS header
+            "seq_count_in_pkts_file": self.l0.ccsds_header.SRC_SEQ_CTR,
+        }
+
+        if level0.LEN == 1:
+            self._process_de_data()
+
+    def append(self, second_l0: DirectEventL0):
+        """Merge an additional direct event packet to this DirectEventL1A class.
+
+        Direct event data can span multiple packets, as marked by the SEQ and LEN
+        attributes. This method will add the next piece of data in the sequence
+        to this data class. The two packets are compared with
+        DirectEventL0.sequence_match_check. If they don't match, the method throws
+        a ValueError.
+
+        If the sequence is broken, the missing sequence numbers are added to
+        missing_seq. Once the last value in the sequence is reached, the data is
+        processed from raw bytes to useful information.
+
+        Parameters
+        ----------
+        second_l0: DirectEventL0
+            Additional L0 packet to add to the DirectEventL1A class
+        """
+        # if SEQ is missing or if the sequence is out of order, do not continue.
+        if not second_l0.SEQ or second_l0.SEQ < self.most_recent_seq:
+            raise ValueError(
+                f"Sequence for direct event L1A is out of order or "
+                f"incorrect. Attempted to append sequence counter "
+                f"{second_l0.SEQ} after {self.most_recent_seq}."
+            )
```

Comment:
Will this raise cause all l1a process to stop? For example, if you have 10 packets and each 5 packets has same SEQ and LEN. You would process first 5 and say this raise happens, does it stop next 5 packets from processing?

If it doesn't stop, then I think we are ok. Otherwise, debug log or something might be helpful.

Reply:
It is my understanding that this shouldn't happen, which is why I raise an error. I think this is something that should be handled by the actual level 1A processing code (which takes in a list of L0 events and produces a list of L1A events) which I haven't written yet, because I think it matches better with the CDF generation and I wanted to get this code reviewed first. 

That being said, this is something I need to ask the GLOWS team about and get a real answer. Possibly I will need to instead track the expected number of packets (which I have) and check off when I receive packets, which would make it so that packets arriving out of order isn't a problem.